components/indexer.py

```diff
@@ -1,56 +1,97 @@
     colour_sensor: rev.ColorSensorV3
-    prox_sensor1: wpilib.DigitalInput
-    prox_sensor2: wpilib.DigitalInput
-    # indexer_piston: wpilib.Solenoid
+    lower_chimney_prox_sensor: wpilib.DigitalInput
+    upper_chimney_prox_sensor: wpilib.DigitalInput
+    cat_flap_piston: wpilib.DoubleSolenoid
 
     is_firing = tunable(False)
-    is_red = tunable(False)
     indexer_speed = tunable(0.5)
-    # Front, back
-    speeds = (0.0, 0.0)
+
+    _tunnel_direction = Direction.OFF
+    _chimney_direction = Direction.OFF
+    _cat_flap_is_open = False
+
+    has_trapped_cargo = tunable(False)
 
     def setup(self) -> None:
-        self.indexer_front_motor.setInverted(True)
-        self.indexer_back_motor.setInverted(False)
+        self.indexer_tunnel_motor.setInverted(True)
+        self.indexer_chimney_motor.setInverted(False)
 
     def execute(self) -> None:
-        self.indexer_front_motor.set(self.speeds[0])
-        self.indexer_back_motor.set(self.speeds[1])
+        self.indexer_tunnel_motor.set(self.indexer_speed * self._tunnel_direction.value)
+        self.indexer_chimney_motor.set(
+            self.indexer_speed * self._chimney_direction.value
+        )
+        if self._cat_flap_is_open:
+            self.cat_flap_piston.set(
+                wpilib.DoubleSolenoid.Value.kForward
+            )  # TODO check direction
+        else:
+            self.cat_flap_piston.set(
+                wpilib.DoubleSolenoid.Value.kReverse
+            )  # TODO check direction
+
+        # Default state is for nothing to be moving and for the cat flap to be down
+        self._tunnel_direction = self._chimney_direction = Indexer.Direction.OFF
+        self._cat_flap_is_open = False
 
     @feedback
-    def has_back(self) -> bool:
-        return not self.prox_sensor1.get() or not self.prox_sensor2.get()
+    def has_cargo_in_chimney(self) -> bool:
+        return (
+            not self.lower_chimney_prox_sensor.get()
+            or not self.upper_chimney_prox_sensor.get()
+        )
 
     @feedback
-    def has_front(self) -> bool:
+    def has_cargo_in_tunnel(self) -> bool:
         return self.colour_sensor.getProximity() > 400
 
     @feedback
-    def is_front_ours(self) -> bool:
+    def are_we_red(self) -> bool:
+        return wpilib.DriverStation.getAlliance() == wpilib.DriverStation.Alliance.kRed
+
+    @feedback
+    def has_opposition_cargo_in_tunnel(self) -> bool:
         raw = self.colour_sensor.getRawColor()
-        return (raw.red > raw.blue) == self.is_red
+        is_red_cargo = raw.red > raw.blue
+        return self.has_cargo_in_tunnel() and is_red_cargo == self.are_we_red()
+
+    @feedback
+    def ready_to_intake(self) -> bool:
+        # We cannot have a cargo in the tunnel, and we can't already have two cargo (one in chimney and one trapped)
+        return not self.has_cargo_in_tunnel() and not (
+            self.has_cargo_in_chimney() and self.has_trapped_cargo
+        )
```

Comment:
Suggesting this as an alternate for readability

```suggestion
        if self.has_cargo_in_tunnel() or  (self.has_cargo_in_chimney() and self.has_trapped_cargo):
             return false

        return true
```

Reply:
I'd personally probably go
```suggestion
        return not (
        	self.has_cargo_in_tunnel()
            or (self.has_cargo_in_chimney() and self.has_trapped_cargo)
        )
```